### ISSUE TITLE
complexCredentials does not accept Promise<ICredentials> or Promise<Credentials> in typescript

### ISSUE THREAD

**Issue by hisham** [awslabs/aws-mobile-appsync-sdk-js#264]
Trying to setup s3 uploads in an angular app with typescript. Setting up appsync with complexObjectCredentials as follows

```
  constructor(private authService: AuthenticationService) {
    this.client = new AWSAppSyncClient({
      url: environment.appSyncGraphQLEndpoint,
      region: environment.appSyncRegion,
      auth: {
        type: AUTH_TYPE.AMAZON_COGNITO_USER_POOLS,
        jwtToken: () => this.amplifyService.auth().currentSession().then( session => {
      return session.getIdToken().getJwtToken();
    })
      },
      complexObjectsCredentials: () => this.amplifyService.auth().getCurrentCredentials(),
      disableOffline: true
    });
  }
```

This results in this typescript error: 

```
error TS2345: Argument of type '{ url: string; region: string; auth: { type: AUTH_TYPE.AMAZON_COGNITO_USER_POOLS; jwtToken: () =>...' is not assignable to parameter of type 'AWSAppSyncClientOptions'.
  Types of property 'complexObjectsCredentials' are incompatible.
    Type '() => Promise<ICredentials>' is not assignable to type '() => Credentials | CredentialsOptions'.
      Type 'Promise<ICredentials>' is not assignable to type 'Credentials | CredentialsOptions'.
        Type 'Promise<ICredentials>' is not assignable to type 'CredentialsOptions'.
          Property 'accessKeyId' is missing in type 'Promise<ICredentials>'.
```

I believe this is because `Promise<Credentials>` or `Promise<ICredentials>` is not added to CredentialsGetter at https://github.com/awslabs/aws-mobile-appsync-sdk-js/blob/f60517639166ceea21ddeb764c66fb6b3d066092/packages/aws-appsync/src/client.ts:

`type CredentialsGetter = () => (Credentials | CredentialsOptions | null) | Credentials | CredentialsOptions | null;`

But you can see when this complex credentials is setup in https://github.com/awslabs/aws-mobile-appsync-sdk-js/blob/ad9dc3cce9986c39bade47a89641868a5cb2641d/packages/aws-appsync/src/link/complex-object-link.ts#L47, promise is supported:

`uploadPromise = Promise.resolve(uploadCredentials)`

You can also see that in this React GraphQl Amplify example app (https://github.com/aws-samples/aws-amplify-graphql/blob/master/photo-client/src/App.js), credentials is being passed as follows: 

`complexObjectsCredentials: () => Auth.currentCredentials()`

Therefore I believe `Promise<Credentials>` and `Promise<ICredentials>` need to be added as a valid return type for CredentialsGetter.

**Comment by rudyhadoux** [comment]
Same blocking problem...

**Comment by hisham** [comment]
The fix for this is now finally available in the latest 1.7.2 package (https://www.npmjs.com/package/aws-appsync). Thank you! Closing this issue.
